source/extensions/clusters/common/logical_host.h

```diff
@@ -9,66 +9,85 @@
-                 // TODO(zyfjeff): Created through metadata shared pool
-                 std::make_shared<const envoy::config::core::v3::Metadata>(lb_endpoint.metadata()),
-                 lb_endpoint.load_balancing_weight().value(), locality_lb_endpoint.locality(),
-                 lb_endpoint.endpoint().health_check_config(), locality_lb_endpoint.priority(),
-                 lb_endpoint.health_status(), time_source),
-        override_transport_socket_options_(override_transport_socket_options) {
-    setAddressList(address_list);
-  }
+      TimeSource& time_source);
 
-  // Set the new address. Updates are typically rare so a R/W lock is used for address updates.
-  // Note that the health check address update requires no lock to be held since it is only
-  // used on the main thread, but we do so anyway since it shouldn't be perf critical and will
-  // future proof the code.
+  /**
+   * Sets new addresses. This can be called dynamically during operation, and
+   * is thread-safe.
+   *
+   * TODO: the health checker only gets the first address in the list and will
+   * not walk the full happy eyeballs list. We should eventually fix this.
+   *
+   * @param address the primary address, also used for health checking
+   * @param address_list alternative addresses; the first of these must be 'address'
+   * @param lb_endpoint the load-balanced endpoint
+   */
   void setNewAddresses(const Network::Address::InstanceConstSharedPtr& address,
-                       const std::vector<Network::Address::InstanceConstSharedPtr>& address_list,
-                       const envoy::config::endpoint::v3::LbEndpoint& lb_endpoint) {
-    const auto& health_check_config = lb_endpoint.endpoint().health_check_config();
-    auto health_check_address = resolveHealthCheckAddress(health_check_config, address);
-    absl::WriterMutexLock lock(&address_lock_);
-    setAddress(address);
-    setAddressList(address_list);
-    /* TODO: the health checker only gets the first address in the list and
-     * will not walk the full happy eyeballs list. We should eventually fix
-     * this. */
-    setHealthCheckAddress(health_check_address);
-  }
+                       const AddressVector& address_list,
+                       const envoy::config::endpoint::v3::LbEndpoint& lb_endpoint);
 
   // Upstream::Host
   CreateConnectionData createConnection(
       Event::Dispatcher& dispatcher, const Network::ConnectionSocket::OptionsSharedPtr& options,
       Network::TransportSocketOptionsConstSharedPtr transport_socket_options) const override;
 
   // Upstream::HostDescription
-  Network::Address::InstanceConstSharedPtr address() const override {
-    absl::ReaderMutexLock lock(&address_lock_);
-    return HostImpl::address();
+  SharedConstAddressVector addressList() const override;
+  Network::Address::InstanceConstSharedPtr address() const override;
+
+private:
+  void setHealthChecker(HealthCheckHostMonitorPtr&& health_checker) override {
```

Comment:
Looking at `setHealthCheckerImpl`, `setOutlierDetector` and `setLastHcPassTime` both implementations seem to do the same. Can they just be implemented once at the base class, or am I missing something?

Reply:
You are right. The problem is the complex relationship where HostImpl and LogicalHostImpl have to inherit from 2 interfaces (Host and HostDescription) and 2 shared implementations (HostDescriptionImplBase and HostImplBase). I couldn't figure out how to share them or the 'resolve' function.

In the case of setHealthCheckerImpl et al, the problem, I think, is that HostDesciprtion does not have the setter API, but does have the getter API. This was pre-existing so I decided not to change it, but maybe this is worth it so that we don't have the new coverage issue.

I think the root

Reply:
I'm going to tackle this one as the alternative is to write new redundant tests to pass coverage. I think it might just add one file to the PR.